src/function/table/arrow.cpp

```diff
@@ -0,0 +1,260 @@
+		} else if (format == "f") {
+			return_types.push_back(LogicalType::FLOAT);
+		} else if (format == "g") {
+			return_types.push_back(LogicalType::DOUBLE);
+		} else if (format == "d:38,0") { // decimal128
+			return_types.push_back(LogicalType::HUGEINT);
+		} else if (format == "u") {
+			return_types.push_back(LogicalType::VARCHAR);
+		} else if (format == "tsn:") {
+			return_types.push_back(LogicalType::TIMESTAMP);
+		} else {
+			throw NotImplementedException("Unsupported Arrow type %s", format);
+		}
+		auto name = string(schema.name);
+		if (name.empty()) {
+			name = string("v") + to_string(col_idx);
+		}
+		names.push_back(name);
+	}
+	data.ReleaseSchema();
+	return move(res);
+}
+
+static void arrow_scan_function(ClientContext &context, vector<Value> &input, DataChunk &output,
+                                FunctionData *dataptr) {
+	auto &data = *((ArrowScanFunctionData *)dataptr);
+
+	if (!data.stream->release) { // no more chunks
+		return;
+	}
+
+	// have we run out of data on the current chunk? move to next one
+	if (data.chunk_offset >= (idx_t)data.current_chunk_root.length) {
+		data.chunk_offset = 0;
+		data.ReleaseArray();
+		if (data.stream->get_next(data.stream, &data.current_chunk_root)) {
+			throw InvalidInputException("arrow_scan: get_next failed(): %s",
+			                            string(data.stream->get_last_error(data.stream)));
+		}
+	}
+
+	// have we run out of chunks? we done
+	if (!data.current_chunk_root.release) {
+		data.stream->release(data.stream);
+		return;
+	}
+
+	if ((idx_t)data.current_chunk_root.n_children != output.column_count()) {
+		throw InvalidInputException("arrow_scan: array column count mismatch");
+	}
+
+	output.SetCardinality(
+	    std::min((int64_t)STANDARD_VECTOR_SIZE, (int64_t)(data.current_chunk_root.length - data.chunk_offset)));
+
+	for (idx_t col_idx = 0; col_idx < output.column_count(); col_idx++) {
+		auto &array = *data.current_chunk_root.children[col_idx];
+		if (!array.release) {
+			throw InvalidInputException("arrow_scan: released array passed");
+		}
+		if (array.length != data.current_chunk_root.length) {
+			throw InvalidInputException("arrow_scan: array length mismatch");
+		}
+		if (array.dictionary) {
+			throw NotImplementedException("arrow_scan: dictionary vectors not supported yet");
+		}
+		if (array.null_count != 0 && array.buffers[0]) {
+			auto &nullmask = FlatVector::Nullmask(output.data[col_idx]);
+
+			auto bit_offset = data.chunk_offset + array.offset;
+			auto n_bitmask_bytes = (output.size() + 8 - 1) / 8;
+
+			if (bit_offset % 8 == 0) {
+				// just memcpy nullmask
+				memcpy(&nullmask, (uint8_t *)array.buffers[0] + bit_offset / 8, n_bitmask_bytes);
+			} else {
+				// need to re-align nullmask :/
+				bitset<STANDARD_VECTOR_SIZE + 8> temp_nullmask;
+				memcpy(&temp_nullmask, (uint8_t *)array.buffers[0] + bit_offset / 8, n_bitmask_bytes + 1);
+
+				temp_nullmask >>= (bit_offset % 8); // why this has to be a right shift is a mystery to me
```

Comment:
In case it's helpful https://github.com/apache/arrow/blob/master/docs/source/format/Columnar.rst#validity-bitmaps